### PR TITLE
#851 manage vaccine item page

### LINKED ISSUE #851: [Vaccine Module][Manage Vaccine Item] - Manage Vaccine Item Page
As per epic: https://github.com/sussol/org-issues/issues/23

#### Description

This page is opened when item is selected from 'Manage Vaccine Stock Page'. Realm.Item is passed with navigator.

VM - Manage Vaccine Item (1)-1

* Table displays all ItemBatches (quantity > 0) for the item
* `Last VVM Pass` column is not needed
* Arrived is from first TB-> transaction date for the ItemBatch
* Table is array of objects (not realm object)
* Apply changes to move to top right and similar style to finalise button (current/historic not needed at this stage)
* Fridge assignment and VVM pass/fail similar to SI
* Breach column can be empty for now
* Create IA when apply changes is pressed
  * If fridge has changed, record in IA
  * Record toggled VVMs in fridge IA and reduce stock for fails
* Ignore dispose button for now

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #851

#### Note: Possibly should close this and do a new PR when I make the page before this one, and other features are in so it can be properly finished, although it's a reasonable size now?

Branch I developed on mainly: [#851-manage-vaccine-page-test](https://github.com/openmsupply/mobile/tree/%23851-manage-vaccine-page-test) (I need a better way to do this .. )

image

Just the skeleton really. Still a few things to do:
- on apply changes (the biggest part, really)
- opening breach modal.
- Navigation, but really is from the parent/prior page component

Other than just creating this component, 
- have updated `IconCell` to also be able to render *just* an icon, rather than *text+icon*. 
- `FinaliseButton` to be able to pass in some text and some font styles.

### CODE REVIEW

File: src/pages/ItemManagePage.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+// Creates a row object for use within this component.
+const createRowObject = (itemBatch, extraData) => {
+  const { id, batch, totalQuantity, location, expiryDate } = itemBatch;
+  return {
+    id: id || generateUUID(),
+    batch,
+    totalQuantity,
+    location: location || { description: 'Unassigned' },
+    expiryDate: typeof expiryDate === 'object' ? expiryDate.toDateString() : expiryDate,
```
Comment (joshxg) [inline src/pages/ItemManagePage.js]: Needs an additional check here for Null

File: src/pages/ItemManagePage.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    if (modalKey) this.setState({ modalKey, isModalOpen: true, currentBatch: itemBatch });
+    else this.setState({ isModalOpen: false, currentBatch: null });
+  };
+
+  /**
+   * RENDER HELPERS
+   */
+  renderCell = (key, itemBatch) => {
+    const { location, vvmStatus } = itemBatch;
+    const modalUpdateProps = { modalKey: key, itemBatch };
```
Comment (andreievg) [inline src/pages/ItemManagePage.js]: currentBatch: itemBatch ? (then don't have to `currentBatch: itemBatch` in onModalUpdate

File: src/pages/ItemManagePage.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  // just update the location of the currentBatch and close the modal.
+  onFridgeSelection = location => {
+    this.updateObject({ location }, { isModalOpen: false });
+  };
+
+  // Called on toggle the vvm toggle bar. If being set to PASS, will just update
+  // the object held in the closure to a PASS VVM status. Otherwise, opens a
+  // modal for a user to enter the doses affected and set the currentBatch.
+  onVvmToggle = currentBatch => ({ newState }) => {
+    if (!newState) return this.setState({ modalKey: 'vvmStatus', isModalOpen: true, currentBatch });
```
Comment (andreievg) [inline src/pages/ItemManagePage.js]: you can use this.onModalUpdateHere, can even replace props with: (modalUpdateProps)

File: src/pages/ItemManagePage.js [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    const usingFridge = vvmStatus !== false && hasFridges;
+    switch (key) {
+      default:
+        return itemBatch[key];
+      case 'location':
+        return (
+          <IconCell
+            text={
+              (hasFridges && 'No Fridges') ||
+              (vvmStatus !== false ? location.description : 'Discarded')
```
Comment (andreievg) [inline src/pages/ItemManagePage.js]: can check for location being null here instead of in createRowObject

Comment (andreievg) [reply src/pages/ItemManagePage.js]: and of course change to: `Unassigned`

File: src/pages/ItemManagePage.js [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+   * RENDER HELPERS
+   */
+  renderCell = (key, itemBatch) => {
+    const { location, vvmStatus } = itemBatch;
+    const modalUpdateProps = { modalKey: key, itemBatch };
+    const hasFridges = this.FRIDGES && this.FRIDGES.length > 0;
+    const usingFridge = vvmStatus !== false && hasFridges;
+    switch (key) {
+      default:
+        return itemBatch[key];
```
Comment (andreievg) [inline src/pages/ItemManagePage.js]: can you use this component please for expiry date (set isEditable to false), it handles having null values

https://github.com/openmsupply/mobile/blob/9dbaf8a9f9115edc2583c3b23c715b81f34b4f8b/src/pages/SupplierInvoicePage.js#L211-L221

File: src/pages/ItemManagePage.js [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+      currentBatch: null,
+    };
+  }
+
+  /**
+   * COMPONENT METHODS
+   */
+  componentDidMount = () => {
+    const { database, item } = this.props;
+    this.FRIDGES = database.objects('Location');
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/ItemManagePage.js]: daaf93d2bd46bd4169ffa09e84e91713ecc3cd7e

File: src/pages/ItemManagePage.js (hunk 1 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: d65bbfd2f87dc63cb5166e904b065200b7e6568b

File: src/pages/ItemManagePage.js (hunk 2 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: cedbc4fa6e8267a90d0e78b1119a00a3088623fe

File: src/pages/ItemManagePage.js (hunk 3 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: nice 694a008cb0e11cd8545e818eaedd4a3dd3ad3c5d

File: src/pages/ItemManagePage.js (hunk 4 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: cf6b43bbf60f0d363d4a2ca12e6416565d756ad9 - Went with "Unnamed?" seems more.. right?

File: src/pages/ItemManagePage.js (hunk 5 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: dad042a1fa82e3be8006fcc4dbb2d64f273f7f05

File: src/pages/ItemManagePage.js (hunk 4 above)
Comment (andreievg) [reply src/pages/ItemManagePage.js]: @joshxg i meant if location is null then 'Unassigned', we will enforce fridge description to be entered and being unique for a store.

File: src/pages/ItemManagePage.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    else this.setState({ isModalOpen: false, currentBatch: null });
+  };
+
+  /**
+   * RENDER HELPERS
+   */
+  renderCell = (key, itemBatch) => {
+    const { location, vvmStatus } = itemBatch;
+    const modalUpdateProps = { modalKey: key, currentBatch: itemBatch };
+    const hasFridges = this.FRIDGES && this.FRIDGES.length > 0;
```
Comment (andreievg) [inline src/pages/ItemManagePage.js]: last thing, this should be in state

File: src/pages/ItemManagePage.js (hunk 4 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: 1b0cdefb68133463cad34740081999d16d504ffb

File: src/pages/ItemManagePage.js (hunk 5 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: 1b0cdefb68133463cad34740081999d16d504ffb

File: src/pages/ItemManagePage.js (hunk 7 above)
Comment (joshxg) [reply src/pages/ItemManagePage.js]: 1b0cdefb68133463cad34740081999d16d504ffb

**Pull request merged**